cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,488 @@
+        my_status = self.xct_solver.runFull()
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = 0 # TODO
+
+        # translate exit status
+        if my_status == 0:
+            if self.xct_solver.hasSolution():
+                self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+            else:
+                self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif my_status == 2:
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN
+        else:
+            raise NotImplementedError(my_status)  # a new status type was introduced, please report on github
+
+        return self.getSolAndObj()
+
+    def solveAll(self, display=None, time_limit=None, solution_limit=None, **kwargs):
+        """
+            Compute all solutions and optionally display the solutions.
+
+            This is the generic implementation, solvers can overwrite this with
+            a more efficient native implementation
+
+            Arguments:
+                - display: either a list of CPMpy expressions, OR a callback function, called with the variables after value-mapping
+                        default/None: nothing displayed
+                - time_limit: stop after this many seconds (default: None)
+                - solution_limit: stop after this many solutions (default: None)
+                - any other keyword argument
+
+            Returns: number of solutions found
+        """
+
+        if not self.has_objective:
+            self.has_objective=True
+            self.xct_solver.init([],[])
+
+        # avoid adding upper bounds when enumerating solutions
+        self.xct_solver.setOption("opt-boundupper",str(0))
+        # set time limit?
+        if time_limit is not None:
+            self.xct_solver.setOption("timeout",str(time_limit))
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            print(kw,val)
+            print(type(self.xct_solver))
+            self.xct_solver.setOption(kw,str(val))
+
+        solsfound = 0
+        while solution_limit == None or solsfound < solution_limit:
+            status = 3
+            while status == 3: # SolveState::INPROCESSED
+                # call the solver, with parameters
+                status = self.xct_solver.runOnce()
+            assert status == 0 or status == 1, "Unexpected status code for Exact."
+            if status == 0: # SolveState::UNSAT
+                break
+            else: # SolveState::SAT
+                solsfound += 1
+                self.getSolAndObj()
+                # TODO: call callback / display?
+                self.xct_solver.invalidateLastSol()
+
+        return solsfound
+
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        assert not is_num(cpm_var)
```

Comment:
are you sure you want to do this?

the 'DirectConstraint' in a pull request is not bothered with calling solver_var on constants, it expects to get a value back if it is a value. Note that you could even do your maximum/str() trick here if you wish...

Reply:
Good question. I was not thinking about outside code calling on this, so I assumed that I could add this as a precondition. I am confused that values and variables are used interchangeably. I guess they function as _terminals_ in the expression tree.

The standard code 
```
if is_num(cpm_var): # shortcut, eases posting constraints
    return cpm_var
```
leads to some `cpm_var` in a direct constraint being replaced, but some `cpm_var` remain, but they actually represent values. So when posting the direct constraint, these `cpm_var` should still be replaced by solver-specific values? This may not be the best approach.

Maybe we should introduce some kind of `solver_val` that the solver can translate to? And adjust `solver_var` accordingly.

Reply:
From a practical point of view, I can add the check, and write a wrapper that adds the assert, only called when translating the constraints (when no values should be passed).

Reply:
Done. DirectConstraint also works now.